packages/next/client/route-loader.ts

```diff
@@ -149,6 +150,38 @@ function appendScript(
   })
 }
 
+function onBuildCallback(cb) {
+  if (process.env.NODE_ENV !== 'development') {
+    cb()
+    return
+  }
+
+  let isComplete = false
+
+  addMessageListener((event) => {
```

Comment:
Is this the best way to subscribe to HMR events? I didn't find a way to unsubscribe, so atm the handler returns if "`isComplete`".

Reply:
Yeah we use `addMessageListener` for the [build indicator here](https://github.com/vercel/next.js/blob/8066e423a7c51e8c670e8328beac91e80da2fc0a/packages/next/client/dev/dev-build-watcher.js#L43-L52) as well.

Let's not move this out into a separate function and keep it inside of `resolvePromiseWithTimeout` and make sure to wrap all dev related code in a `if (process.env.NODE_ENV === 'development') {}` check so that it can be dead-code eliminated to prevent impacting production bundle sizes. We can use `require` to load the `addMessageListener` inside of the `if` check as well.